sklearn/linear_model/sgd_fast.pyx

```diff
@@ -705,13 +756,31 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
                 infinity = True
                 break
 
-            if tol > -INFINITY and sumloss > previous_loss - tol * n_samples:
-                if verbose:
-                    with gil:
-                        print("Convergence after %d epochs took %.2f seconds"
-                              % (epoch + 1, time() - t_start))
-                break
-            previous_loss = sumloss
+            # evaluate the score on the validation set
+            if early_stopping:
+                with gil:
+                    score = estimator._validation_score(weights, intercept)
+                if tol > -INFINITY and score < previous_score + tol:
+                    no_improvement_count += 1
```

Comment:
What is exactly happening if I have a sequence of: no improvement, improvement, no improvement. I have the impression that we do not reinitialize the improvement counter in this case, isn't it (I might miss the line thought :))

Reply:
Good point, I fixed it.